python/ray/train/mosaic/_mosaic_utils.py

```diff
@@ -0,0 +1,225 @@
+            eval_dataset, batch_size, labels
+        )
+    else:
+        eval_torch_iterable = None
+
+    return train_torch_iterable, eval_torch_iterable
+
+
+class RayLogger(LoggerDestination):
+    """A logger to relay information logged by composer models to ray.
+
+    This logger allows utilizing all necessary logging and logged data handling provided
+    by the Composer library. All the logged information is saved in the data dictionary
+    every time a new information is logged, but to reduce unnecessary reporting, the
+    most up-to-date logged information is reported as metrics every batch checkpoint and
+    epoch checkpoint (see Composer's Event module for more details).
+
+    Because ray's metric dataframe will not include new keys that is reported after the
+    very first report call, any logged information with the keys not included in the
+    first batch checkpoint would not be retrievable after training. In other words, if
+    the log level is greater than `LogLevel.BATCH` for some data, they would not be
+    present in `Result.metrics_dataframe`. To allow preserving those information, the
+    user can provide keys to be always included in the reported data by using `keys`
+    argument in the constructor. For `MosaicTrainer`, use
+    `trainer_init_config['log_keys']` to populate these keys.
+
+    Args:
+        log_level: the granuality to log data. The default value is ``LogLevel.BATCH``
+    """
+
+    def __init__(
+        self, log_level: Union[str, int, LogLevel] = LogLevel.BATCH, keys=[]
+    ) -> None:
+        self.log_level = LogLevel(log_level)
+        self.data = {}
+        for key in keys:
+            self.data[key] = None
+
+    def log_data(self, state: State, log_level: LogLevel, data: Dict[str, Any]):
+        if log_level > self.log_level:
+            # the logged metric is more verbose than what we want to record.
+            return
+        self.data.update(data.items())
+        for key, val in self.data.items():
+            if isinstance(val, torch.Tensor):
+                self.data[key] = val.item()
+
+    def batch_checkpoint(self, state: State, logger: Logger) -> None:
+        del logger  # unused
+        session.report(self.data)
+
+    def epoch_checkpoint(self, state: State, logger: Logger) -> None:
+        del logger  # unused
+        session.report(self.data)
+
+
+class RayTrainReportCallback(CheckpointSaver):
+    """A callback that wraps Composer's ``CheckpointSaver``.
+
+    This class is used to wrap each Composer ``CheckpointSaver`` to be used in Composer
+    trainer. The main role of this callback is to report the paths of the checkpoints
+    saved by the Composer ``CheckpointSaver`` it wraps. In addition, when the training
+    ends, (either with or without exception) the last checkpoint and the list of all
+    the checkpoints that have been saved and the list of Composer InMemoryLogger are
+    reported.
```

Comment:
Shouldn't we instead create checkpoints from those paths (`Checkpoint.from_directory`) and report those? I am not sure how they are going to be transferred to other nodes in the current logic. It seems like it will only work on a single node to me.